### PR TITLE
Add testing for concurrently loading/using/destroying the same key

### PR DESCRIPTION
**Pull request by Ryan-Everett-arm**
## Description
There are two parts to this PR. The first part is repetitive work which changes the key smoke tests to support the possibility that the key may be destroyed at any time. The second part is adding tests for various key behaviours.

We first add a new parameter `key_destroyable` to the `exercise_X` functions. When `key_destroyable == 0` the behaviour is the same as before these changes, when `key_destroyable == 1` the smoke tests will ignore any valid failures caused by the key suddenly not existing; unexpected failures like corruption will be reported.

The reason we are doing this is that it provides us a framework which we can use to test that we abide by our key access rules set in `psa-thread-safety.md`:

> * FULL: the slot contains a key, and any thread is able to use the key after registering as a reader.
> * PENDING_DELETION: the key within the slot has been destroyed or marked for destruction, but at least one thread is still registered as a reader. No thread can register to read this slot. The slot must not be wiped until the last reader de-registers, wiping the slot by calling `psa_wipe_key_slot`.

By running `mbedtls_test_psa_exercise_key` on the key in one thread, and concurrently destroying the key in another, TSan/internal errors will spot if these properties are being violated.

1. If an operation reads the slot contents without being registered - there will be a read/write conflict and the thread will try to read NULL data, causing an error.
2. If a thread wipes the slot before everyone has deregistered - the same will occur.

This PR adds testing for the following:
1. Exactly one thread succeeds when multiple threads attempt to import the same persistent key.
2. Exactly one thread succeeds when multiple threads attempt to destroy the same key (#8912).
3. The correct error is returned when trying to access a free slot when no slot exists.
4. Any thread can use a key as soon as it is loaded by another thread.
5. Threads do not access the slot material when not registered as readers.
6. Threads do not wipe the contents of the slot while other threads are registered as readers (ensures destruction in use doesn't cause corruption).

**Disclaimer**: this doesn't test every API call or every code path. Exercise key can do a lot, but it doesn't cover all types of operation or all API calls. We also only do not cover all methods of loading a key, or all types of keys. Future work should add more functionality to exercise key, add more test data to this function, and add more test functions which use the other key management functions.
## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** not required
- [x] **tests** yes

### CODE REVIEW

**Comment by Ryan-Everett-arm** [comment]
In its current form this PR does not handle the potential INSUFFICIENT_MEMORY failure which can occur in exercise key when lots of threads attempt to load into a key slot. It shouldn't be merged before this is changed.

File: tests/suites/test_suite_psa_crypto.function [hunk 1]
```diff
@@ -1338,6 +1338,127 @@ exit:
...
+/* Attempt to import the key in ctx. This handles any valid error codes
+ * and reports an error for any invalid codes. This function also insures
+ * that once imported by some thread, all threads can use the key. */
+void *thread_import_key(void *ctx)
+{
+    mbedtls_svc_key_id_t returned_key_id;
+    same_key_context *skc = (struct same_key_context *) ctx;
+    psa_key_attributes_t got_attributes = PSA_KEY_ATTRIBUTES_INIT;
+
+    /* Import the key, exactly one thread must succceed. */
```
Comment (valeriosetti) [inline tests/suites/test_suite_psa_crypto.function:1364]: Typo: succeed

File: tests/suites/test_suite_psa_crypto.function [hunk 2]
```diff
@@ -1338,6 +1338,127 @@ exit:
...
+    psa_status_t status = psa_import_key(skc->attributes, skc->data->x,
+                                         skc->data->len, &returned_key_id);
+    switch (status) {
+        case PSA_SUCCESS:
+            if (mbedtls_mutex_lock(&skc->key_loaded_mutex) == 0) {
+                if (skc->key_loaded) {
+                    mbedtls_mutex_unlock(&skc->key_loaded_mutex);
+                    /* More than one thread has succeeded, report a failure. */
+                    TEST_EQUAL(skc->key_loaded, 0);
```
Comment (valeriosetti) [inline tests/suites/test_suite_psa_crypto.function:1373]: Why not directly `TEST_FAIL()` in this case? I think it would be more explicit, but I might be missing something.
If you accept this change, then there is a very similar case also in `thread_use_and_destroy_key()` below

**Review by valeriosetti — COMMENTED** [review]
I did a first pass to get more familiar with the PR and it looks almost OK to me. I only found very few comments. I will do a second pass tomorrow.

**Comment by Ryan-Everett-arm** [comment]
> In its current form this PR does not handle the potential INSUFFICIENT_MEMORY failure which can occur in exercise key when lots of threads attempt to load into a key slot. It shouldn't be merged before this is changed.

Upon revisiting this PR I realize that I was mistaken here. My concern was of two (what I thought where) possible scenarios:

1. `exercise_key_derivation_key` calls `key_derivation_output_key` - It doesn't, it only ever calls `key_derivation_output_bytes`.
2. A scenario could arise within a set of calls to `exercise_key(k)` where multiple slots are reserved for the same persistent key. - The only key ever used in this situation is `k`, and our locking policy in `psa_get_and_lock_key_slot` means that there can **never** be a situation where two threads calling `psa_get_and_lock_key_slot_X` attempt to load the same persistent key into multiple slots concurrently.

To summarize: no extra resources are ever required when calling `psa_exercise_key` concurrently on the same key, so there is no risk of `INSUFFICIENT_MEMORY` occuring in any sensible test. I will add a comment to the definition of `exercise_key` which documents that this function cannot be called concurrently with more unique persistent keys than there are free key slots.

File: tests/suites/test_suite_psa_crypto.function (hunk 2 above)
Comment (Ryan-Everett-arm) [reply tests/suites/test_suite_psa_crypto.function]: I wasn't aware of `TEST_FAIL()`, good suggestion!

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Review by valeriosetti — APPROVED** [review]
Thanks for addressing my comments. LGTM :)
